x-pack/plugins/osquery/public/shared_components/osquery_action/index.tsx

```diff
@@ -5,79 +5,135 @@
-        const indexPattern = await indexPatterns.find('.fleet-agents');
-
-        searchSource.setField('index', indexPattern[0]);
-        searchSource.setField('filter', [
-          {
-            meta: {
-              alias: null,
-              disabled: false,
-              negate: false,
-              key: 'local_metadata.host.id',
-              value: hostId,
-            },
-            query: {
-              match_phrase: {
-                'local_metadata.host.id': hostId,
-              },
-            },
-          },
-          {
-            meta: {
-              alias: null,
-              disabled: false,
-              negate: false,
-              key: 'active',
-              value: 'true',
-            },
-            query: {
-              match_phrase: {
-                active: 'true',
-              },
-            },
-          },
-        ]);
-
-        const response = await searchSource.fetch$().toPromise();
-
-        if (response.rawResponse.hits.hits.length && response.rawResponse.hits.hits[0]._id) {
-          setAgentId(response.rawResponse.hits.hits[0]._id);
+const OsqueryActionComponent: React.FC<OsqueryActionProps> = ({ metadata }) => {
+  const permissions = useKibana().services.application.capabilities.osquery;
+  const agentId = metadata?.info?.agent?.id ?? undefined;
+  const {
+    data: agentData,
+    isFetched: agentFetched,
+    isError: agentError,
+    isLoading: agentLoading,
+  } = useAgentDetails({
+    agentId,
+    silent: true,
+    skip: !agentId,
+  });
+  const {
+    data: agentPolicyData,
+    isFetched: policyFetched,
+    isError: policyError,
+    isLoading: policyLoading,
+  } = useAgentPolicy({
+    policyId: agentData?.item?.policy_id,
+    skip: !agentData,
+    silent: true,
+  });
+
+  const osqueryAvailable = useMemo(() => {
+    if (policyError) return false;
+
+    const osqueryPackageInstalled = find(agentPolicyData?.package_policies, [
+      'package.name',
+      OSQUERY_INTEGRATION_NAME,
+    ]);
+    return osqueryPackageInstalled?.enabled;
+  }, [agentPolicyData?.package_policies, policyError]);
+
+  if (!(permissions.runSavedQueries || permissions.writeLiveQueries)) {
+    return (
+      <EuiEmptyPrompt
+        icon={<OsqueryIcon />}
+        title={<h2>Permissions denied</h2>}
+        titleSize="xs"
+        body={
+          <p>
+            To access this page, ask your administrator for <EuiCode>osquery</EuiCode> Kibana
+            privileges.
+          </p>
         }
-      };
+      />
+    );
+  }
 
-      findAgent();
-    }
-  });
+  if (!agentFetched) {
+    return <EuiLoadingContent lines={10} />;
+  }
 
-  if (!agentId) {
+  if (!agentId || (agentFetched && agentError)) {
+    return (
+      <EuiEmptyPrompt
+        icon={<OsqueryIcon />}
+        title={<h2>Osquery is not available</h2>}
+        titleSize="xs"
+        body={
+          <p>
+            An Elastic Agent is not installed on this host. To run queries on the host, install
```

Comment:
In reading this again, "on this host" is repetitive, and I suggest this change:

```suggestion
            An Elastic Agent is not installed on this host. To run queries, install
```